File: C++/Leetcode 438.cpp
#include <stdio.h>
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> findAnagrams(string s2, string s1) {
            // Reverse Analogy

        if (s1.length() > s2.length())
        {
            return {};
        }

        int length = s1.length();
        vector<int>answer;

        unordered_map<int,int> maps1;
        unordered_map<int,int> maps2;

        for (int i = 0; i < s1.length(); i++)
        {
            maps1[s1[i]]++;
        }

        int st_index = 0;

        for (st_index = 0; st_index < length; st_index++)
        {
            maps2[s2[st_index]]++;
        }

        if (maps1 == maps2)
        {
            // answer++;
            answer.push_back(0);
        }

        int last_index = 0;

        while (st_index < s2.length())
        {
            maps2[s2[st_index]]++;
            maps2[s2[last_index]]--;

            if (!maps2[s2[last_index]])
            {
                maps2.erase(s2[last_index]);
            }

            st_index++;
            last_index++;
            // cout << st_index  - last_index << endl;
            if (maps2 == maps1)
            {
                // cout << last_index << "   " << st_index << endl;
                answer.push_back(last_index);
                // return true;
            }
        }
        return answer;
    }
};

File: Python/Leetcode 438.py
class Solution:
    def findAnagrams(self, s2: str, s1: str) -> List[int]:
        if len(s1) > len(s2):
            return []
        
        Dict1 = dict()
        Dict2 = dict()

        for char in s1:
            Dict1[char] = Dict1.get(char,0) + 1

        end_index = 0
        answer_list = []
        
        while end_index < len(s1):
            Dict2[s2[end_index]] = Dict2.get(s2[end_index],0) + 1
            end_index = end_index + 1
        
        if Dict1 == Dict2:
            answer_list.append(0)
        
        st_index = 0


        while end_index < len(s2):
            Dict2[s2[end_index]] = Dict2.get(s2[end_index],0) + 1
            Dict2[s2[st_index]] = Dict2.get(s2[st_index],0) - 1

            if Dict2[s2[st_index]] == 0:
                Dict2.pop(s2[st_index])

            end_index = end_index + 1
            st_index = st_index + 1

            if Dict2 == Dict1:
                answer_list.append(st_index)
        
        return answer_list
